enclave_wrangler/objects_api.py

```diff
@@ -1024,6 +1044,157 @@ def get_projects(verbose=False) -> List[Dict]:
+     This applies to the 'for cset in cset_objs' part. In order to run update_objects_in_db() on all cset_objs at once,
+     I should also update the audit_query to run on multiple objs at once. Also, it may be useful that updated rows are
+     sorted 1 at a time in the audit table. So that with default sorting / or sorting explicitly by the update
+     timestamp, you see the before/after for each cset, rather than all of the csets before the update, then all the
+     csets after the update, with all the timestamps in the before/after sets being equal due to the same audit_query
+     for each set.
+    """
+    conn = con if con else get_db_connection()
+    cset_objs = cset_objs if isinstance(cset_objs, list) else [cset_objs]
+
+    if set_missing_fields_to_null:
+        cset_objs = cset_objs_set_missing_fields_to_null(cset_objs, schema)
+
+    audit_query = """INSERT INTO code_sets_audit
+        SELECT *, CURRENT_TIMESTAMP AS update_timestamp
+        FROM code_sets
+        WHERE codeset_id = cset_id;"""
+    for cset in cset_objs:
+        cset_id = str(cset.get('codesetId', cset.get('codeset_id')))
+        run_sql(con, audit_query.replace('cset_id', cset_id))  # backup to audit table: original
+        update_objects_in_db(con, 'OMOPConceptSet', [cset])
+        run_sql(con, audit_query.replace('cset_id', cset_id))  # backup to audit table: updated
+    if not con:
+        conn.close()
+
+
+def _get_cset_row_diffs_by_field(list1: List[Dict], list2: List[Dict]) -> Dict[int, Dict[str, Tuple[Any, Any]]]:
+    """Get differences between two sets of rows of the same type of codeset, e.g. 1 from DB and 1 from Enclave.
+
+    Result: Dict where keys are codeset_id, and values are inner dictionaries, which have as keys, any key in each
+     object where the values differ between their appearance in , and values are a tuple of 2 values, where 1 is the
+     val from the first list (e.g. DB) and the other is from the corresponding entry from the other list (e.g. Enclave).
+    """
+    report = {}
+    enclave_dict = {obj.get('codeset_id', obj.get('codesetId')): obj for obj in list2}
+    for db_obj in list1:
+        codeset_id = db_obj['codeset_id', db_obj.get('codesetId')]
+        if codeset_id in enclave_dict:
+            enclave_obj = enclave_dict[codeset_id]
+            # Compare all keys and store differences
+            differences = {}
+            all_keys = set(db_obj.keys()) | set(enclave_obj.keys())
+            for key in all_keys:
+                db_value = db_obj.get(key)
+                enclave_value = enclave_obj.get(key)
+                if db_value != enclave_value:
+                    differences[key] = (db_value, enclave_value)
+            if differences:
+                report[codeset_id] = differences
+    return report
+
+
+def clean_improper_finalized_draft_metadata(check_even_if_still_draft=False, verbose=False, codeset_ids: List[int]=None):
```

Comment:
## Fixed `[DRAFT]` in finalized cset titles
Checked any csets where title `LIKE` `[DRAFT]`, and made sure their metadata was up-to-date. This resulted many fixes where we did not update the cset when the draft was finalized. There were also some cases of csets that were still drafts but that have had metadata changes too.